File: src/shared/modelShapes.ts
import type { ConfigControl, SessionControls } from './transcript';
import { groupModels } from './models';

// Per-agent memory of the dependent controls (effort, Fast, other model parameters) each model came with, keyed by the
// model option id. Agents reshape these per model (Devin 3000.11.3: SWE-2 has no `speed` and only medium / high / max
// effort), and only a live `set_config_option` reveals the new shape; the history editor switches models locally, so it
// reads a model's parameters from here instead of keeping the previous model's.
export type ModelShapes = Record<string, ConfigControl[]>;

export function isModelControl(c: Pick<ConfigControl, 'id' | 'category'>): boolean {
  return c.category === 'model' || (!c.category && c.id === 'model');
}

// Structure only: a value change (another effort picked) is not a new shape and must not rewrite the shared prefs file
const structure = (controls: ConfigControl[]) => JSON.stringify(controls.map(c => [c.id, c.category, c.type, c.options.map(o => o.id)]));

// The map with the current model's shape recorded, or undefined when there is no model or its shape is already known
export function learnShape(shapes: ModelShapes | undefined, controls: SessionControls): ModelShapes | undefined {
  const model = controls.options.find(isModelControl);
  if (!model?.value) return;
  const shape = controls.options.filter(c => !isModelControl(c));
  const known = shapes?.[model.value];
  if (known && structure(known) === structure(shape)) return;
  return { ...shapes, [model.value]: shape };
}

// A model's remembered shape; an unseen id borrows one from its family (Devin's effort and Fusion variants share the parameters)
export function shapeFor(shapes: ModelShapes | undefined, model: ConfigControl, value: string): ConfigControl[] | undefined {
  if (!shapes) return;
  if (shapes[value]) return shapes[value];
  const family = groupModels(model.options).find(f => f.variants.some(v => v.id === value));
  const sibling = family?.variants.find(v => shapes[v.id]);
  return sibling && shapes[sibling.id];
}

File: rust/crates/acpira-shared/src/model_shapes.rs
//! Per-agent memory of the dependent controls each model came with (mirror of src/shared/modelShapes.ts)

use std::collections::BTreeMap;

use crate::models::group_models;
use crate::transcript::{ConfigControl, SessionControls};

pub type ModelShapes = BTreeMap<String, Vec<ConfigControl>>;

pub fn is_model_control(c: &ConfigControl) -> bool {
  c.category.as_deref() == Some("model") || (c.category.is_none() && c.id == "model")
}

// Structure only: a value change is not a new shape
fn structure(controls: &[ConfigControl]) -> Vec<(&str, Option<&str>, bool, Vec<&str>)> {
  controls
    .iter()
    .map(|c| (c.id.as_str(), c.category.as_deref(), c.kind.is_some(), c.options.iter().map(|o| o.id.as_str()).collect()))
    .collect()
}

/// The map with the current model's shape recorded, or None when there is no model or its shape is already known
pub fn learn_shape(shapes: Option<&ModelShapes>, controls: &SessionControls) -> Option<ModelShapes> {
  let model = controls.options.iter().find(|c| is_model_control(c))?;
  let value = model.value.as_ref()?;
  let shape: Vec<ConfigControl> = controls.options.iter().filter(|c| !is_model_control(c)).cloned().collect();
  if let Some(known) = shapes.and_then(|s| s.get(value))
    && structure(known) == structure(&shape)
  {
    return None;
  }
  let mut next = shapes.cloned().unwrap_or_default();
  next.insert(value.clone(), shape);
  Some(next)
}

/// A model's remembered shape; an unseen id borrows one from its family
pub fn shape_for<'a>(shapes: Option<&'a ModelShapes>, model: &ConfigControl, value: &str) -> Option<&'a Vec<ConfigControl>> {
  let shapes = shapes?;
  if let Some(s) = shapes.get(value) {
    return Some(s);
  }
  let families = group_models(&model.options);
  let family = families.iter().find(|f| f.variants.iter().any(|v| v.id == value))?;
  let sibling = family.variants.iter().find(|v| shapes.contains_key(&v.id))?;
  shapes.get(&sibling.id)
}
